tests/solver_tester.py

```diff
@@ -0,0 +1,111 @@
+import logging
+from abc import ABC
+
+import torch
+
+from dynamiqs.options import Options
+
+from .system import System
+
+
+class SolverTester(ABC):
+    def _test_batching(self, options: Options, system: System):
+        """Test the batching of `H` and `y0`, and the returned object sizes."""
+        m, n = system._state_shape
+        n_exp_ops = len(system.exp_ops)
+        b_H = len(system.H_batched)
+        b_y0 = len(system.y0_batched)
+        num_t_save = 11
+        t_save = system.t_save(num_t_save)
+
+        run = lambda H, y0: system._run(H, y0, t_save, options)
+
+        # no batching
+        result = run(system.H, system.y0)
+        assert result.y_save.shape == (num_t_save, m, n)
+        assert result.exp_save.shape == (n_exp_ops, num_t_save)
+
+        # batched H
+        result = run(system.H_batched, system.y0)
+        assert result.y_save.shape == (b_H, num_t_save, m, n)
+        assert result.exp_save.shape == (b_H, n_exp_ops, num_t_save)
+
+        # batched y0
+        result = run(system.H, system.y0_batched)
+        assert result.y_save.shape == (b_y0, num_t_save, m, n)
+        assert result.exp_save.shape == (b_y0, n_exp_ops, num_t_save)
+
+        # batched H and y0
+        result = run(system.H_batched, system.y0_batched)
+        assert result.y_save.shape == (b_H, b_y0, num_t_save, m, n)
+        assert result.exp_save.shape == (b_H, b_y0, n_exp_ops, num_t_save)
+
+    def test_batching(self):
+        pass
+
+    def _test_correctness(
+        self,
+        options: Options,
+        system: System,
+        *,
+        num_t_save: int,
+        y_save_norm_atol: float = 1e-2,
+        exp_save_rtol: float = 1e-2,
+        exp_save_atol: float = 1e-2,
+    ):
+        t_save = system.t_save(num_t_save)
+        result = system.run(t_save, options)
+
+        # === test y_save
+        errs = torch.linalg.norm(result.y_save - system.states(t_save), dim=(-2, -1))
+        assert torch.all(errs <= y_save_norm_atol)
+
+        # === test exp_save
+        assert torch.allclose(
+            result.exp_save,
+            system.expects(t_save),
+            rtol=exp_save_rtol,
+            atol=exp_save_atol,
+        )
+
+    def test_correctness(self):
+        pass
+
+    def _test_gradient(
```

Comment:
Can you separate this function in two subfunctions for readability? A first one `_test_gradient_y` and a second one `_test_gradient_exp`, and then wrap them both inside your current `_test_gradient`.

Reply:
I'm not sure it's a good idea as (1) both part of the function share common variables (`system`, `t_save` and `result`) that we'd have to pass around, and (2) we don't want these subfunctions to be called individually. I guess the comments with `===` are enough to separate both parts no?